core/src/main/python/model_diff.py

```diff
@@ -0,0 +1,681 @@
+            found_in_past_dictionary = self._in_model(self.past_dict, itm)
+            debug('DBUEG: found_in_past_dictionary %s', found_in_past_dictionary)
+            if not found_in_past_dictionary:
+                break
+            else:
+                # check whether it is in the forbidden list
+                if self.in_forbidden_list(itm):
+                    compare_msgs(('WLSDPLY-05303', itm))
+                    return FATAL_MODEL_CHANGES
+
+        # if there is a shape change
+        # return 2 ?
+        if has_topology and not found_in_past_dictionary:
+            compare_msgs.add(('WLSDPLY-05302', itm))
+            return FATAL_MODEL_CHANGES
+
+        if found_in_past_dictionary:
+            return SAFE_ONLINE_UPDATE
+
+        # allow new additions for anything ??
+        return SAFE_ONLINE_UPDATE
+
+    def _in_model(self, dictionary, keylist):
+        """
+        check whether the keys is in the dictionary
+        :param dictionary dictonary to check
+        :param keylist  dot separted key list
+
+        return 1 if it is in model
+               0 if it is not in model
+        """
+        debug('DBEUG: in model keylist=%s dictionary %s', keylist, dictionary)
+
+        splitted=keylist.split('|')
+        n=len(splitted)
+        i=0
+
+        # loop through the keys and use it to walk the dictionary
+        # if it can walk down 3 levels, safely assume it is in the
+        # dictionary, otherwise it is a total new addition
+
+        for i in range(0, n):
+            if dictionary.has_key(splitted[i]):
+                if isinstance(dictionary[splitted[i]], dict):
+                    dictionary = dictionary[splitted[i]]
+                continue
+            else:
+                break
+
+        if i > 2:
+            return 1
+
+        return 0
+
+    def in_forbidden_list(self, itm):
+        forbidden_list = [ '.ListenPort', '.ListenAddress' ]
+        for forbidden in forbidden_list:
+            if itm.endswith(forbidden):
+                return 1
+        return 0
+
+    def get_final_changed_model(self):
+        """
+        Return the changed model.
+        """
+        return self.final_changed_model
+
+
+class ModelFileDiffer:
+    """
+      This is the main driver for the caller.  It compares two model files whether they are json or yaml format.
+    """
+    def __init__(self, current_dict, past_dict, output_dir=None):
+        self.current_dict_file = current_dict
+        self.past_dict_file = past_dict
+        self.output_dir = output_dir
+
+    def get_dictionary(self, file):
+        """
+        Retrieve the python dictionary from file
+        :param file:  disk file containing the python dictionary
+        :return:  python dictionary
+        """
+        true = True
+        false = False
+        fh = open(file, 'r')
+        content = fh.read()
+        return eval(content)
+
+
+    def write_dictionary_to_json_file(self, dictionary, writer, indent=''):
```

Comment:
could PythonToJson in json_translator.py be used for this? might save some duplication.

Reply:
Yes, fixing it